### PR TITLE
Support multiple schedules per plan

### LINKED ISSUE #23: Allow multiple backup plans
Have a question? Please checkout our [Slack Community](https://slack.cloudposse.com) or visit our [Slack Archive](https://archive.sweetops.com/). 

[Slack Community](https://slack.cloudposse.com)

## Describe the Feature

It would be nice to be able to create multiple backup plans. Otherwise we have to change the `name` so there aren't iam role name conflicts.

## Expected Behavior

Allow multiple backup plans and allow `selection_tags` to use a `list(map(string))` instead of `list(string)`

## Describe Ideal Solution

Create multiple backup plans with their own selection tags in a single instantiation of this module

## Alternatives Considered

Multiple instantiations with different `name` set

## Additional Context

N/A

### PR DESCRIPTION
**Pull request by nightspotlight**
## what
* Add support for multiple backup schedules per backup plan
* Simplify configuration in case multiple schedules per backup plan are necessary
* Maintain backwards compatibility

## why
* One has to have multiple instances of this module to have more than one backup schedule, each with a different value for name
* In my case, I need to have three schedules in a backup plan: daily, weekly, monthly

## references
* Closes #23

### CODE REVIEW

**Comment by nightspotlight** [comment]
@nitrocode , what do you think?

**Comment by nitrocode** [comment]
Thanks for the contribution. I like the fact that you're creating multiple rules per backup plan.

The way your logic is structured shows that if var.rules is set, then it will use var.rules, and if anything is unset in var.rules, it will default to the current input.

i.e.

```hcl
              cold_storage_after = lookup(rule.value, "copy_action_cold_storage_after", var.copy_action_cold_storage_after)
```

This effectively turns all the current inputs into "global" inputs and I'm not sure if that's what we want here.

It may be better to set all of the 3rd arguments to `null` instead and move the older inputs into a `variables-deprecated.tf`, then use a `local` to construct the `rules` depending on if the deprecated inputs or the new `rules` input is used. We'd also need to add a small doc in `docs/` on how to migrate.

Please also update the `examples/complete` to use the new `rules` input.

**Comment by nightspotlight** [comment]
@nitrocode ,

> The way your logic is structured shows that if var.rules is set, then it will use var.rules, and if anything is unset in var.rules, it will default to the current input.

> This effectively turns all the current inputs into "global" inputs and I'm not sure if that's what we want here.

that's exactly what I was trying to achieve - to maintain compatibility in case some users don't have or have loose version constraint for this module. for that reason, I wouldn't like to deprecate existing variables. if you insist on deprecating, I'm ready to make another change with a migration instruction.

**Comment by nitrocode** [comment]
I think you're trying to achieve backward compatibility but what you end up with is setting `global` inputs.

Yes, I think deprecating would be best.

See this as an example https://github.com/cloudposse/terraform-aws-msk-apache-kafka-cluster/pull/44

**Comment by nightspotlight** [comment]
> I think you're trying to achieve backward compatibility but what you end up with is setting `global` inputs.

Yes, I'd thought that defaulting to "global" inputs would have an added benefit of allowing users to set shared settings across all rules, e.g.:
```hcl
  rules = [
    {
      name     = "daily"
      schedule = "cron(0 10 * * ? *)"
    },
    {
      name     = "monthly"
      schedule = "cron(0 12 1 * ? *)"
    }
  ]
  start_window       = 60
  completion_window  = 120
  cold_storage_after = 30
  delete_after       = 180
```

But as you suggested, I have deprecated those inputs and added a migration guide. Please review.

File: main.tf [hunk 1]
```diff
@@ -8,6 +8,24 @@ locals {
   vault_name       = coalesce(var.vault_name, module.this.id)
   vault_id         = join("", local.vault_enabled ? aws_backup_vault.default.*.id : data.aws_backup_vault.existing.*.id)
   vault_arn        = join("", local.vault_enabled ? aws_backup_vault.default.*.arn : data.aws_backup_vault.existing.*.arn)
+  compat_rule = {
```
Comment (nitrocode) [inline main.tf:11]: what does `compat` mean ? Is this a typo for `compact` ?

Comment (nitrocode) [reply main.tf:11]: Or is it short for compatible ?

**Review by nitrocode — APPROVED** [review]
Looks good to me. Nice job!

I requested one more review before we merge just to make sure we didn't miss anything.

**Comment by nitrocode** [comment]
> Yes, I'd thought that defaulting to "global" inputs would have an added benefit of allowing users to set shared settings across all rules, e.g.:

If that was the intention, I think maybe we want to add that in a separate PR. Perhaps we can even call it `global_rule` to make it clear what it will do. I did not like the previous change as it would have been a surprise if anyone was to use both the original inputs and the new `rules` input.

File: main.tf (hunk 1 above)
Comment (nightspotlight) [reply main.tf:11]: Yes, "compat" is for "compatibility" (I've seen this usage in https://cgit.freebsd.org/src/tree/sys/compat/linux). I wanted to make it apparent that this record is for backwards compatibility and is likely not staying in the code forever. Let me know if you'd like me to change it.

**Review by Benbentwo — APPROVED** [review]

**Pull request merged**